azure/bedrock api integration

sameelarif opened the pull request:
# why
Need to support Azure/Bedrock and all other AI SDK providers that have more than just an `apiKey` parameter.

# what changed
Allowed free input of `modelClientOptions` and added logic for authenticating with AWS for Bedrock usage.

# test plan

sameelarif committed:
allow free modelclientoptions

sameelarif committed:
Merge branch 'main' into sameel/stg-692-azurebedrock-api-integration-serialize

sameelarif committed:
Merge branch 'main' into sameel/stg-692-azurebedrock-api-integration-serialize

sameelarif committed:
change zod ver to working build

sameelarif committed:
send client options on every request

miguelg719 commented on lib/api.ts:
```diff
@@ -261,6 +265,12 @@ export class StagehandAPI {
+    // Add modelClientOptions as a header if provided
+    if (modelClientOptions) {
+      defaultHeaders["x-model-client-options"] =
+        JSON.stringify(modelClientOptions);
```
why send it as a header? modelClientOptions already gets sent in the payload. A stringified json in a header is probably not the move

filip-michalsky committed:
test bedrock file

filip-michalsky committed:
add azure test file

sameelarif committed:
fix bedrock test

sameelarif committed:
Merge branch 'main' into sameel/stg-692-azurebedrock-api-integration-serialize

sameelarif committed:
Update pnpm-lock.yaml

miguelg719 committed:
update evals cli docs (#1096)

# why

# what changed

# test plan

sameelarif committed:
better modelclientoption api handling

sameelarif committed:
dont override region

sameelarif committed:
fix bedrock example

filip-michalsky commented on examples/example-bedrock.ts:
```diff
@@ -0,0 +1,101 @@
+  // Initialize Stagehand with AWS Bedrock
+  const stagehand = new Stagehand({
+    env: "BROWSERBASE",
+    modelName: "bedrock/us.anthropic.claude-sonnet-4-20250514-v1:0",
```
You can't hard code the model name unless you very clearly and in detail add to the readme how a user requests that particular model name from the AWS dashboard @sameelarif

filip-michalsky replied:
I couldnt run the example bc of that

Kylejeong2 committed:
adding support for new claude 4.5 sonnet agent model (#1099)

# why

anthropic released a new sota computer use model

# what changed

added claude-sonnet-4-5-20250929 as a model to the list

# test plan

ran evals

miguelg719 commented on lib/api.ts:
```diff
@@ -191,6 +185,12 @@ export class StagehandAPI {
+    this.logger({
+      category: "execute",
+      message: `Executing ${method} with args: ${JSON.stringify(args)}`,
+      level: 1,
```
maybe level 2

sameelarif committed:
lint

sameelarif committed:
read aws creds from client options obj

miguelg719 commented on lib/StagehandPage.ts:
```diff
@@ -747,6 +747,7 @@ ${scriptContent} \
             const result = await this.api.act({
               ...observeResult,
               frameId: this.rootFrameId,
+              modelClientOptions: this.stagehand["modelClientOptions"],
```
is this required?

miguelg719 commented on lib/utils.ts:
```diff
@@ -493,6 +494,153 @@ export function loadApiKeyFromEnv(
+    message:
+      "No Amazon Bedrock authentication credentials found. Please provide credentials via modelClientOptions (accessKeyId/secretAccessKey or bearerToken) or environment variables (AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_BEARER_TOKEN_BEDROCK)",
+    level: 0,
+  });
```
should we throw here

sameelarif replied:
It would throw once they make an LLM call. We let people use stagehand without an LLM so I figured we just print a warning for now. Down to change it tho

tkattkat committed:
properly convert custom / mcp tools to anthropic cua format (#1103)

Why

Custom AI SDK tools and MCP integrations weren't working properly with
Anthropic CUA - parameters were empty {} and tools weren't tracked.

What Changed


- Convert Zod schemas to JSON Schema before sending to Anthropic (using
zodToJsonSchema)
- Track custom tool calls in the actions array
- Silence "Unknown tool name" warnings for custom tools

Test Plan

Tested with examples file. 

Parameters passed correctly ({"city":"San Francisco"} instead of {})
Custom tools execute and appear in actions array
No warnings

miguelg719 committed:
Add current date and page url to agent context (#1102)

# why
To improve context

# what changed
Added current page and url to the system prompt

# test plan

miguelg719 committed:
Additional agent logging (#1104)

# why
To inform the user throughout the agent execution process

# what changed
Added logs to tool calls, and on the stagehand agent handler

# test plan
- [x] tested locally

miguelg719 committed:
fix system prompt

miguelg719 committed:
remove dup log

miguelg719 committed:
pass modelClientOptions for stagehand agent

sameelarif commented on lib/StagehandPage.ts:
```diff
@@ -844,7 +846,10 @@ ${scriptContent} \
-          result = await this.api.extract<T>({ frameId: this.rootFrameId });
+          result = await this.api.extract<T>({
+            frameId: this.rootFrameId,
+            modelClientOptions: this.stagehand["modelClientOptions"],
```
Yes because otherwise it doesn't get sent to the API. We need this param on all api calls now

sameelarif committed:
Merge branch 'main' into sameel/stg-692-azurebedrock-api-integration-serialize
